### PR TITLE
[SCRIPT] Add ESIM for text matching

### PR DESCRIPTION
**Pull request by kenjewu**
## Description ##

Build an Enhancing LSTM model for Natural Language Inference. 
The model does not currently contain the TreeLSTM mentioned in the original paper. Training scripts and tests are in progress.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: scripts/esim/esim.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    ndense_units : int
+        Number of hidden units in dense layer
+    nclasses : int
+        Number of categories
+    drop_out : int
+        Dropout prob
+    """
+
+    def __init__(self, nwords, nword_dims, nhiddens, ndense_units,
+                 nclasses, drop_out, **kwargs):
```
Comment (hhexiy) [inline scripts/esim/esim.py]: Suggestion on variable naming: 
- `nwords` -> `vocab_size` (more self-explanatory)
- `drop_out` -> `dropout=0.` (convention)
- `nhiddens` -> `nhidden_units` (for consistency with `ndense_units`)

File: scripts/esim/esim.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+"""
+
+__all__ = ['ESIM']
+
+from mxnet.gluon import nn, rnn
+
+
+class ESIM(nn.HybridBlock):
+    """"Enhanced LSTM for Natural Language Inference" Qian Chen,
+    Xiaodan Zhu, Zhenhua Ling, Si Wei, Hui Jiang, Diana Inkpen. ACL (2017)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: Add paper URL.

File: scripts/esim/esim.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                                nn.Dense(nclasses))
+
+            self.avg_pool = nn.GlobalAvgPool1D()
+            self.max_pool = nn.GlobalMaxPool1D()
+
+    def _soft_attention_align(self, F, x1, x2, mask1, mask2):
+        # attention shape: (batch, seq_len, seq_len)
+        attention = F.batch_dot(x1, x2, transpose_b=True)
+
+        weight1 = F.softmax(attention + F.expand_dims(mask2, axis=1), axis=-1)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: @szha can add `masked_softmax` as a util function.

Should we mask `weight1` to make sure that masked dims in `weight1` is zero?

File: scripts/esim/esim.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        return F.concat(mul, sub, dim=-1)
+
+    def _apply_multiple(self, F, x):
+        p1 = F.squeeze(self.avg_pool(x), axis=-1)
+        p2 = F.squeeze(self.max_pool(x), axis=-1)
+
+        return F.concat(p1, p2, dim=-1)
+
+    def hybrid_forward(self, F, x1, x2, mask1, mask2):  # pylint: disable=arguments-differ
+        # x1_embed x2_embed shape: (batch, seq_len, nword_dims)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: (batch, x1_seq_len, x2_seq_len)

File: scripts/esim/esim.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                                nn.ELU(),
+                                nn.BatchNorm(axis=-1),
+                                nn.Dropout(drop_out),
+                                nn.Dense(nclasses))
+
+            self.avg_pool = nn.GlobalAvgPool1D()
+            self.max_pool = nn.GlobalMaxPool1D()
+
+    def _soft_attention_align(self, F, x1, x2, mask1, mask2):
+        # attention shape: (batch, seq_len, seq_len)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: # attention shape: (batch, x1_seq_len, x2_seq_len)

File: scripts/esim/esim.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        attention = F.batch_dot(x1, x2, transpose_b=True)
+
+        weight1 = F.softmax(attention + F.expand_dims(mask2, axis=1), axis=-1)
+        x1_align = F.batch_dot(weight1, x2)
+        weight2 = F.softmax(attention + F.expand_dims(mask1, axis=1), axis=-1)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: I think attention needs to be transposed here.

File: scripts/esim/esim.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    def _submul(self, F, x1, x2):
+        mul = F.multiply(x1, x2)
+        sub = F.subtract(x1, x2)
+
+        return F.concat(mul, sub, dim=-1)
+
+    def _apply_multiple(self, F, x):
+        p1 = F.squeeze(self.avg_pool(x), axis=-1)
+        p2 = F.squeeze(self.max_pool(x), axis=-1)
+
```
Comment (hhexiy) [inline scripts/esim/esim.py]: Assuming shape of `x` is `N x T x C`, Pooling should happen across `T`, i.e. `axis=1`.

File: scripts/esim/esim.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        weight1 = F.softmax(attention + F.expand_dims(mask2, axis=1), axis=-1)
+        x1_align = F.batch_dot(weight1, x2)
+        weight2 = F.softmax(attention + F.expand_dims(mask1, axis=1), axis=-1)
+        x2_align = F.batch_dot(weight2, x1)
+
+        return x1_align, x2_align
+
+    def _submul(self, F, x1, x2):
+        mul = F.multiply(x1, x2)
+        sub = F.subtract(x1, x2)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: You can do 
```
x1 * x2
x1 - x2
```

File: scripts/esim/esim.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+        return F.concat(mul, sub, dim=-1)
+
+    def _apply_multiple(self, F, x):
```
Comment (hhexiy) [inline scripts/esim/esim.py]: Rename to `_pooling`? `_apply_multiple` is not self-explanatory.

File: scripts/esim/esim.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                                                        mask1, mask2)
+
+        # compose
+        x1_combined = F.concat(x1_lstm_encode, x1_algin,
+                               self._submul(F, x1_lstm_encode, x1_algin), dim=-1)
+        x2_combined = F.concat(x2_lstm_encode, x2_algin,
+                               self._submul(F, x2_lstm_encode, x2_algin), dim=-1)
+
+        x1_compose = self.lstm_encoder2(x1_combined)
+        x2_compose = self.lstm_encoder2(x2_combined)
```
Comment (hhexiy) [inline scripts/esim/esim.py]: Should project `x1_combined` and `x2_combined` back to 300D before sending to LSTM.

File: scripts/esim/esim.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                 nclasses, drop_out, **kwargs):
+        super(ESIM, self).__init__(**kwargs)
+        with self.name_scope():
+            self.embedding_encoder = nn.Embedding(nwords, nword_dims)
+            self.batch_norm = nn.BatchNorm(axis=-1)
+            self.lstm_encoder1 = rnn.LSTM(nhiddens, bidirectional=True)
+            self.lstm_encoder2 = rnn.LSTM(nhiddens, bidirectional=True)
+
+            self.fc_encoder = nn.HybridSequential()
+            self.fc_encoder.add(nn.BatchNorm(axis=-1),
```
Comment (hhexiy) [inline scripts/esim/esim.py]: In my experiment using a standard 1 layer FFN is good enough. The current `fc_encoder` gives NaN results somehow.

**Comment by kenjewu** [comment]
@hhexiy 
Thanks for the comment. The model code is okay on a competition task. But it has not been run on public data. Testing and calibration work is in progress. I will check and test it according to your proposal, and ping you after the work is completed.

File: scripts/esim/esim.py (hunk 11 above)
Comment (kenjewu) [reply scripts/esim/esim.py]: The reason for this is that I previously mistyped the softmax axis of weight2. Many bugs have been fixed in the new submission based on your comments and original papers. I will do a full test on the data mentioned in the paper later. Thank you very much.

File: scripts/esim/esim.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                                nn.Dropout(dropout),
+                                nn.Dense(nhidden_units, activation='relu', flatten=False))
+
+            self.lstm_encoder2 = rnn.LSTM(nhidden_units, bidirectional=True)
+
+            self.fc_encoder = nn.HybridSequential()
+            self.fc_encoder.add(nn.BatchNorm(axis=-1),
+                                nn.Dropout(dropout),
+                                nn.Dense(ndense_units),
+                                nn.ELU(),
```
Comment (fierceX) [inline scripts/esim/esim.py]: In `mxnet=1.4.1`, nn.ELU() has a bug. For details, please refer to https://github.com/apache/incubator-mxnet/pull/14673

**Comment by mli** [comment]
Job PR-689/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-689/2/index.html

**Comment by mli** [comment]
Job PR-689/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-689/3/index.html

**Comment by mli** [comment]
Job PR-689/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-689/4/index.html

**Comment by mli** [comment]
Job PR-689/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-689/5/index.html

**Comment by mli** [comment]
Job PR-689/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-689/6/index.html

**Comment by szha** [comment]
Some items left:
- [ ] report performance and upload log to dmlc/web-data
- [ ] update model zoo page to include the description of the model, the command, and link to the log

**Comment by eric-haibin-lin** [comment]
@szha move the remaining items to github issue to track ?
